sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -32,209 +32,325 @@ import org.apache.spark.util.Utils
+   * Adds the expression to this data structure, including its children recursively.
    */
-  private def updateExprInMap(
+  def addExprTree(expr: Expression): Unit = {
+    if (supportedExpression(expr)) {
+      updateMapWithExprTree(equivalenceMap, expr, 1, 0)
+    }
+  }
+
+  private def supportedExpression(expr: Expression) = {
+    !expr.exists {
+      // `LambdaVariable` is usually used as a loop variable, which can't be evaluated ahead of the
+      // loop. So we can't evaluate sub-expressions containing `LambdaVariable` at the beginning.
+      case _: LambdaVariable => true
+
+      // `PlanExpression` wraps query plan. To compare query plans of `PlanExpression` on executor,
+      // can cause error like NPE.
+      case _: PlanExpression[_] => Utils.isInRunningSparkTask
+
+      case _ => false
+    }
+  }
+
+  private def updateMapWithExprTree(
+      map: mutable.HashMap[ExpressionEquals, ExpressionStats],
       expr: Expression,
+      evalCount: Int,
+      conditionalEvalCount: Double): Unit = {
+    if (!expr.isInstanceOf[LeafExpression]) {
+      updateMapWithExpr(map, expr, evalCount, conditionalEvalCount)
+
+      expr match {
+        // CodegenFallback's children will not be used to generate code (call eval() instead)
+        case _: CodegenFallback =>
+
+        case c: CaseWhen =>
+          // Let's consider `CaseWhen(Seq((w1, t1), (w2, t2), (w3, t3), ... (wn, tn)), Some(e))`
+          // example and use `Wn`, `Tn` and `E` notations for the local equivalence maps built from
+          // `wn`, `tn` and `e` expressions respectively.
+          //
+          // Let's try to build a local equivalence map of the above `CaseWhen` example and then add
+          // that local map to `map`.
+          //
+          // We know that `w1` is surely evaluated so `W1` should be part of the local map.
+          // We also know that based on the result of `w1` either `t1` or `w2` is evaluated so the
+          // "intersection" between `T1` and `W2` should be also part of the local map.
+          // Please note that "intersection" might not describe well the operation that we need
+          // between `T1` and `W2`. It is an intersection in terms of surely evaluated
+          // subexpressions (`evalCount`) between `T1` and `W2` but it is also kind of an union
+          // between conditionally evaluated subexpressions (`conditionalEvalCount`). See the
+          // details in `intersectMapWithMap()`.
+          // So the local map can be calculated as `W1 | (T1 & W2)` so far, where `|` and `&` mean
+          // the "union" and "intersection" of equivalence maps.
+          // But we can continue the previous logic further because if `w2` is evaluated, then based
+          // on the result of `w2` either `t2` or `w3` is also evaluated.
+          // So eventually the local equivalence map can be calculated as
+          // `W1 | (T1 & W2 | T2 & (W3 | T3 & ... & (Wn | Tn & E)))`.
```

Comment:
Is there a missing parenthesis here? I'm trying to understand the order of operations once you get to `T2`.
Is it `W1 | T1 & (W2 | T2 & (W3 | ...`
assuming normal higher precedence of `&`

Reply:
\+ I've changed the new config in: https://github.com/apache/spark/pull/41677/commits/38a0996aa1dbf2471e6a5d4fa4dc031995b9beb0 and from now the default 0 value has the same effect as `spark.sql.subexpressionElimination.conditionals.enabled` in your PR.